### PR TITLE
feat: DeleteRange Phase 2 — memtable read path and fragmenter

### PR DESCRIPTION
**Pull request by ben1009**
## Summary

Implements Phase 2 of the DeleteRange feature ([RFC 010](https://github.com/ben1009/toy-kv-engine/blob/main/rfcs/010-delete-range.md)).

### Fragmenter (`range_tombstone.rs`)
- `fragment_range()` — sweep-line algorithm converting raw tombstones to non-overlapping fragments. Uses `BTreeMap<u64, u32>` refcount multiset for correct handling of duplicate timestamps.
- `merge_fragment_lists()` — O(F log F) sweep-line merge of fragment lists from multiple sources with timestamp union and coalescing.
- `RangeTombstoneIterator` — binary-search based lookup over sorted fragments for O(log F + log T) per-key coverage checks.
- 5 new unit tests (duplicate ts, iterator boundaries).

### Memtable (`mem_table.rs`)
- `ImmutableRangeTombstoneSet` — wraps raw `SkipMap` with `OnceLock<Arc<[RangeTombstoneFragment]>>` for lazy fragment cache on frozen memtables.
- `freeze_range_tombstones()` — interior mutability via `OnceLock` (works through `&self` since memtable is already in `Arc`).

### Read Path (`lsm_storage.rs`)
- `get()` — checks range tombstones across all memtables after finding point candidate; hides key if `value_ts <= range_ts`. Pre-computes `newest_memtable_range_ts` once for both memtable and SST checks.
- `get_with_ts()` — added range tombstone check for transaction point lookups (was missing, snapshot isolation violation).
- `get_with_kind_inner()` — added range tombstone check for CAS operations.
- `scan_inner()` — builds merged `RangeTombstoneIterator` from active + immutable memtables, passes to `LsmIterator`.

### Iterator (`lsm_iterator.rs`)
- `skip_tombstones()` — decodes user key from memcomparable form only when range tombstones are present; checks coverage after finding visible point version.

### Tests (`tests/memtable.rs`)
- 8 new integration tests: get/scan with range tombstones, get_with_ts visibility, flush guard, multi-version key interaction, cross-layer resurrection.

## Verification
- [x] 512 tests pass
- [x] 0 clippy warnings
- [x] 3-round adversarial review completed and issues fixed

## Summary by CodeRabbit

* **New Features**
  * Range tombstones now filter key visibility for both point lookups and scans.
  * MVCC reads now account for range-tombstone coverage using the read timestamp.
  * Scan operations efficiently skip keys hidden by `[start, end)` deletions, including correct half-open end behavior.
* **Bug Fixes**
  * Reads now consistently short-circuit covered versions to ensure expected deletion semantics.
* **Tests**
  * Added extensive integration coverage for range-tombstone behavior across reads, freezes into immutable state, and timestamp/visibility edge cases.

### CODE REVIEW

File: kv-engine/src/lsm_storage.rs [hunk 1]
```diff
@@ -1657,20 +1705,71 @@ impl LsmStorageInner {
...
             self.lookup_sst_raw(state, encoded, bloom_hash, mvcc_read_ts)?
         {
+            if let Some(rts) = mvcc_read_ts {
+                let range_ts = self.newest_memtable_range_ts(state, key, rts);
+                if range_ts.is_some_and(|rt| value_ts <= rt) {
+                    return Ok((None, KvKind::Inline));
+                }
+            }
             return Ok((val, kind));
         }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs:1737]: Fixed. Hoisted range_ts computation before both if-let blocks so it is computed once and reused.

File: kv-engine/src/lsm_storage.rs [hunk 2]
```diff
@@ -1657,20 +1705,71 @@ impl LsmStorageInner {
...
+            if let Some(imm) = m.imm_range_tombstones() {
+                let ts = imm.newest_covering_ts(user_key, read_ts);
+                if ts.is_some() && (best_ts.is_none() || ts > best_ts) {
+                    best_ts = ts;
+                }
+            }
+        }
+
+        best_ts
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Applied. Refactored newest_memtable_range_ts to use fold over immutable memtables with Option::max for cleaner iteration.

File: kv-engine/src/range_tombstone.rs [hunk 3]
```diff
@@ -192,6 +193,268 @@ impl Default for RangeTombstoneSet {
...
+    // two tombstones sharing the same ts don't lose coverage when one ends
+    // before the other.
+    let mut active: BTreeMap<u64, u32> = BTreeMap::new();
+    // Pre-index: for each endpoint, which tombstones start / end here.
+    let mut starts_at: Vec<Vec<(usize, u64)>> = vec![Vec::new(); endpoints.len()];
+    let mut ends_at: Vec<Vec<(usize, u64)>> = vec![Vec::new(); endpoints.len()];
+
+    for (idx, (start, end, ts)) in tombstones.iter().enumerate() {
+        let start_pos = endpoints.binary_search(start).unwrap();
+        let end_pos = endpoints.binary_search(end).unwrap();
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs]: Fixed. Replaced .unwrap() with .expect() documenting the invariant that endpoints must exist in collected endpoints.

File: kv-engine/src/range_tombstone.rs [hunk 4]
```diff
@@ -192,6 +193,268 @@ impl Default for RangeTombstoneSet {
...
+    // Store (fragment_idx, covering_ts_slice_idx) but since we need all ts
+    // from a fragment, store the fragment index and expand later.
+    let mut starts_at: Vec<Vec<usize>> = vec![Vec::new(); endpoints.len()];
+    let mut ends_at: Vec<Vec<usize>> = vec![Vec::new(); endpoints.len()];
+
+    for (idx, f) in all.iter().enumerate() {
+        let start_pos = endpoints.binary_search(&f.start).unwrap();
+        let end_pos = endpoints.binary_search(&f.end).unwrap();
+        starts_at[start_pos].push(idx);
+        ends_at[end_pos].push(idx);
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:395]: Fixed. Replaced .unwrap() with .expect() documenting the invariant that endpoints must exist in collected endpoints.

**Comment by ben1009** [comment]
All review comments have been addressed and resolved:

1. **range_ts double computation**: Hoisted computation before both if-let blocks in `get_with_kind_inner()`
2. **fold for newest_memtable_range_ts**: Refactored to use `fold` with `Option::max`
3. **unwrap → expect**: Replaced `.unwrap()` with `.expect()` in `fragment_range` and `merge_fragment_lists`
4. **Dedup binary search**: Delegated `ImmutableRangeTombstoneSet::newest_covering_ts` to `RangeTombstoneIterator`
5. **Snapshot test**: Added `get_with_ts` assertion in `test_get_snapshot_before_delete_range`

Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/range_tombstone.rs [hunk 5]
```diff
@@ -356,4 +627,299 @@ mod tests {
...
+            end: Bytes::from_static(b"m"),
+            covering_ts: vec![10],
+        }];
+        let frags2 = vec![RangeTombstoneFragment {
+            start: Bytes::from_static(b"m"),
+            end: Bytes::from_static(b"z"),
+            covering_ts: vec![10],
+        }];
+
+        let merged = merge_fragment_lists(&[frags1, frags2]);
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs]: Fixed. Updated test call sites for new merge_fragment_lists signature that accepts &[&[RangeTombstoneFragment]].

File: kv-engine/src/range_tombstone.rs [hunk 6]
```diff
@@ -356,4 +627,299 @@ mod tests {
...
+            end: Bytes::from_static(b"z"),
+            covering_ts: vec![10],
+        }];
+        let frags2 = vec![RangeTombstoneFragment {
+            start: Bytes::from_static(b"m"),
+            end: Bytes::from_static(b"p"),
+            covering_ts: vec![20],
+        }];
+
+        let merged = merge_fragment_lists(&[frags1, frags2]);
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs]: Fixed. Updated test call sites for new merge_fragment_lists signature that accepts &[&[RangeTombstoneFragment]].

File: kv-engine/src/range_tombstone.rs [hunk 7]
```diff
@@ -356,4 +627,299 @@ mod tests {
...
+            start: Bytes::from_static(b"a"),
+            end: Bytes::from_static(b"m"),
+            covering_ts: vec![10],
+        }];
+        let frags2 = vec![RangeTombstoneFragment {
+            start: Bytes::from_static(b"m"),
+            end: Bytes::from_static(b"z"),
+            covering_ts: vec![20],
+        }];
+
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:783]: Fixed. Updated test call sites for new merge_fragment_lists signature that accepts &[&[RangeTombstoneFragment]].

File: kv-engine/src/range_tombstone.rs [hunk 8]
```diff
@@ -192,6 +193,276 @@ impl Default for RangeTombstoneSet {
...
+/// Merge multiple sorted fragment lists into a single non-overlapping list.
+///
+/// Each input list must be sorted by `start` and non-overlapping within itself.
+/// The output is sorted by `start`, non-overlapping, with `covering_ts` being
+/// the union of all input sources for each span. Adjacent fragments with
+/// identical `covering_ts` are coalesced.
+pub fn merge_fragment_lists(lists: &[Vec<RangeTombstoneFragment>]) -> Vec<RangeTombstoneFragment> {
+    // Collect all fragments and sort by start.
+    let all: Vec<&RangeTombstoneFragment> = lists.iter().flatten().collect();
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs]: Fixed. Changed merge_fragment_lists to accept &[&[RangeTombstoneFragment]] instead of &[Vec<...>] to avoid cloning immutable memtable fragments on every scan.

File: kv-engine/src/lsm_storage.rs [hunk 9]
```diff
@@ -1600,7 +1620,35 @@ impl LsmStorageInner {
...
+                None
+            } else {
+                let merged = crate::range_tombstone::merge_fragment_lists(&[all_fragments]);
+                Some(crate::range_tombstone::RangeTombstoneIterator::new(
+                    Arc::from(merged),
+                ))
+            }
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Fixed. Collects Arc references from immutable memtables and passes borrowed slices to merge_fragment_lists, avoiding all cloning.

File: kv-engine/src/lsm_storage.rs [hunk 10]
```diff
@@ -1184,17 +1186,27 @@ impl LsmStorageInner {
...
-        if let Some((value, kind, found_key)) =
+        if let Some((value, kind, found_key, value_ts)) =
             self.lookup_memtable(&state, encoded, bloom_hash, mvcc_read_ts)?
         {
-            return self.resolve_value(&found_key, value, kind);
+            if range_ts.is_some_and(|rt| value_ts <= rt) {
+                // Covered by range tombstone — fall through to SST lookup.
+            } else {
+                return self.resolve_value(&found_key, value, kind);
+            }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs:1201]: Fixed. When the memtable value is covered by a range tombstone, returns None immediately — any SST version is strictly older and also covered.

File: kv-engine/src/lsm_storage.rs [hunk 11]
```diff
@@ -1381,14 +1393,22 @@ impl LsmStorageInner {
...
+        let range_ts = self.newest_memtable_range_ts(&state, key, read_ts);
+        if let Some((value, kind, found_key, value_ts)) =
             self.lookup_memtable(&state, &lookup_key, bloom_hash, Some(read_ts))?
         {
-            return self.resolve_value(&found_key, value, kind);
+            if range_ts.is_some_and(|rt| value_ts <= rt) {
+                // Covered — fall through to SST lookup.
+            } else {
+                return self.resolve_value(&found_key, value, kind);
+            }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs:1403]: Fixed. When the memtable value is covered by a range tombstone, returns None immediately — any SST version is strictly older and also covered.

File: kv-engine/src/lsm_storage.rs [hunk 12]
```diff
@@ -1657,20 +1705,56 @@ impl LsmStorageInner {
...
+        let range_ts = mvcc_read_ts.and_then(|rts| self.newest_memtable_range_ts(state, key, rts));
+        if let Some((val, kind, _key, value_ts)) =
             self.lookup_memtable(state, encoded, bloom_hash, mvcc_read_ts)?
         {
-            return Ok((val, kind));
+            if range_ts.is_some_and(|rt| value_ts <= rt) {
+                // Covered — fall through to SST lookup.
+            } else {
+                return Ok((val, kind));
+            }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs:1726]: Fixed. When the memtable value is covered by a range tombstone, returns None immediately — any SST version is strictly older and also covered.

**Comment by ben1009** [comment]
All review comments have been addressed and resolved:

1. **merge_fragment_lists signature**: Changed to accept `&[&[RangeTombstoneFragment]]` to avoid cloning immutable memtable fragments on every scan
2. **Early return on cover**: When memtable value is covered by range tombstone, returns None immediately (any SST version is strictly older and also covered)
3. **Test call sites**: Updated for new merge_fragment_lists signature

Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/range_tombstone.rs [hunk 13]
```diff
@@ -192,6 +193,276 @@ impl Default for RangeTombstoneSet {
...
+/// Each input list must be sorted by `start` and non-overlapping within itself.
+/// The output is sorted by `start`, non-overlapping, with `covering_ts` being
+/// the union of all input sources for each span. Adjacent fragments with
+/// identical `covering_ts` are coalesced.
+pub fn merge_fragment_lists(lists: &[&[RangeTombstoneFragment]]) -> Vec<RangeTombstoneFragment> {
+    // Collect all fragments and sort by start.
+    let all: Vec<&RangeTombstoneFragment> = lists.iter().copied().flatten().collect();
+    if all.is_empty() {
+        return Vec::new();
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:370]: Fixed. Added short-circuit for empty input and single list — returns immediately without sweep-line overhead.

File: kv-engine/src/lsm_iterator.rs [hunk 14]
```diff
@@ -69,15 +75,27 @@ impl LsmIterator {
...
             // and reuse it in both the read_ts filtering and tombstone skip.
             encoded_user_key.clear();
+            decoded_user_key.clear();
             if TS_ENABLED {
                 encoded_user_key.extend_from_slice(iter.key().encoded_user_key());
+                // Decode to raw user key only when range tombstones are present.
+                // Range tombstones store raw user keys, not memcomparable form.
+                if range_ts_iter.is_some() {
+                    crate::key::decode_user_key_into(&encoded_user_key, &mut decoded_user_key);
+                }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_iterator.rs:112]: Not fixed. KeySlice does not have a decode_user_key_into method — only decode_user_key() which allocates a new Vec per call. The current approach reuses the decoded_user_key buffer across iterations, which is more efficient for the scan hot path.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/lsm_iterator.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_iterator.rs:112]: Fixed. Now uses iter.key().decode_user_key_into() with buffer clearing before each decode.

File: kv-engine/src/lsm_storage.rs [hunk 15]
```diff
@@ -1600,7 +1620,44 @@ impl LsmStorageInner {
...
+                crate::range_tombstone::fragment_range(state.memtable.range_tombstones().raw())
+            } else {
+                Vec::new()
+            };
+            // Keep Arc alive for the duration of the merge.
+            let imm_arcs: Vec<Arc<[crate::range_tombstone::RangeTombstoneFragment]>> = state
+                .imm_memtables
+                .iter()
+                .filter_map(|m| m.imm_range_tombstones().map(|imm| imm.fragments()))
+                .collect();
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Fixed. fragments() now returns &Arc, with explicit .clone() at call sites that need owned Arc.

File: kv-engine/src/mem_table.rs [hunk 16]
```diff
@@ -23,6 +26,52 @@ use crate::{
...
+    /// Delegates to `RangeTombstoneIterator::newest_covering_ts` to avoid
+    /// duplicating the binary-search logic.
+    pub fn newest_covering_ts(&self, user_key: &[u8], read_ts: u64) -> Option<u64> {
+        self.iter().newest_covering_ts(user_key, read_ts)
+    }
+
+    /// Build a `RangeTombstoneIterator` for scan-path use.
+    pub fn iter(&self) -> RangeTombstoneIterator {
+        RangeTombstoneIterator::new(self.fragments())
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/mem_table.rs]: Fixed. fragments() now returns &Arc<[RangeTombstoneFragment]> to avoid atomic refcount operations on the hot point-lookup path. Callers that need owned Arc use explicit .clone().

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/mem_table.rs [hunk 17]
```diff
@@ -23,6 +26,51 @@ use crate::{
...
+            Arc::from(frags)
+        })
+    }
+
+    /// Find the newest covering tombstone timestamp for `user_key` at `read_ts`.
+    ///
+    /// Delegates to `RangeTombstoneIterator::newest_covering_ts` to avoid
+    /// duplicating the binary-search logic.
+    pub fn newest_covering_ts(&self, user_key: &[u8], read_ts: u64) -> Option<u64> {
+        self.iter().newest_covering_ts(user_key, read_ts)
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/mem_table.rs]: Fixed. Inlined the binary search directly on the fragment slice returned by fragments() to avoid Arc clone on every point lookup.

File: kv-engine/src/range_tombstone.rs [hunk 18]
```diff
@@ -192,6 +193,280 @@ impl Default for RangeTombstoneSet {
...
+
+        let covering_ts: Vec<u64> = active.keys().copied().collect();
+
+        fragments.push(RangeTombstoneFragment {
+            start: frag_start,
+            end: frag_end,
+            covering_ts,
+        });
+    }
+
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:351]: Not fixed. The current O(F log F) sweep-line is already efficient for the expected workload. The flat event-based approach would save some Vec allocations but adds code complexity with diminishing returns — the endpoint count is typically small (2x fragment count).

File: kv-engine/src/range_tombstone.rs [hunk 19]
```diff
@@ -192,6 +193,280 @@ impl Default for RangeTombstoneSet {
...
+        }
+
+        let covering_ts: Vec<u64> = active.keys().copied().collect();
+        fragments.push(RangeTombstoneFragment {
+            start: frag_start,
+            end: frag_end,
+            covering_ts,
+        });
+    }
+
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:452]: Not fixed. The current O(F log F) sweep-line is already efficient for the expected workload. The flat event-based approach would save some Vec allocations but adds code complexity with diminishing returns — the endpoint count is typically small (2x fragment count).

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

**Comment by ben1009** [comment]
Fixed. Borrowing directly from `state.imm_memtables` in `scan_inner` — removed the intermediate `imm_arcs` Vec and Arc clones entirely. The `state` reference lives long enough to cover the entire scan.

**Comment by ben1009** [comment]
Fixed. Added `tmp_encoded_key` and `tmp_decoded_key` as `Vec<u8>` fields on `LsmIterator`, initialized once in `new()` and reused across all `skip_tombstones` calls via `.clear()`. Eliminates per-call heap allocations on the hot scan path.

**Comment by ben1009** [comment]
Fixed. `next()` now passes `&mut self.tmp_encoded_key` and `&mut self.tmp_decoded_key` to `skip_tombstones`, avoiding heap allocations on every iteration.

**Comment by ben1009** [comment]
Fixed. `skip_tombstones` now accepts `encoded_user_key: &mut Vec<u8>` and `decoded_user_key: &mut Vec<u8>` parameters, reusing the caller buffers instead of allocating new ones each invocation.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

**Comment by ben1009** [comment]
Fixed. Replaced `if let Some(ts) = read_ts` with `let ts = read_ts.expect("MVCC must be enabled for snapshot visibility test")` — the assertion now always executes, and the test will fail loudly if MVCC is not configured.

**Comment by ben1009** [comment]
Not fixed. As noted in round 4, the current O(F log F) sweep-line is already efficient for the expected workload. The event-based approach would save some Vec allocations but adds code complexity with diminishing returns for typical tombstone counts.

**Comment by ben1009** [comment]
Not fixed. Same reasoning as above — the current sweep-line approach is already efficient and the added complexity of an event-based refactor does not justify the marginal improvement.

**Comment by ben1009** [comment]
Fixed. Removed the redundant `decoded_user_key.clear()` — it was already called unconditionally at the top of the loop on line 105.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

**Comment by ben1009** [comment]
Not fixed. Same reasoning as rounds 4 and 6 — the current O(F log F) sweep-line is already efficient. The event-based refactor adds complexity with diminishing returns for typical tombstone counts.

**Comment by ben1009** [comment]
Not fixed. Same reasoning — the current sweep-line is efficient enough and the marginal improvement does not justify the added complexity.

**Comment by ben1009** [comment]
Not fixed. The lazy `OnceLock::get_or_init` approach is idiomatic and the first-read latency is negligible for typical tombstone counts. Eagerly computing fragments during `freeze_range_tombstones` would shift cost to the write-critical path.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/range_tombstone.rs [hunk 20]
```diff
@@ -192,6 +193,280 @@ impl Default for RangeTombstoneSet {
...
+        fragments.push(RangeTombstoneFragment {
+            start: frag_start,
+            end: frag_end,
+            covering_ts,
+        });
+    }
+
+    fragments
+}
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:353]: Not fixed. The current O(F log F) sweep-line is already efficient for the expected workload. The event-based refactor adds complexity with diminishing returns for typical tombstone counts.

File: kv-engine/src/range_tombstone.rs [hunk 21]
```diff
@@ -192,6 +193,280 @@ impl Default for RangeTombstoneSet {
...
+        let covering_ts: Vec<u64> = active.keys().copied().collect();
+        fragments.push(RangeTombstoneFragment {
+            start: frag_start,
+            end: frag_end,
+            covering_ts,
+        });
+    }
+
+    fragments
+}
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:454]: Not fixed. Same reasoning — the current sweep-line is efficient enough and the marginal improvement does not justify the added complexity.

File: kv-engine/src/mem_table.rs [hunk 22]
```diff
@@ -627,6 +701,25 @@ impl MemTable {
...
+    /// Uses `OnceLock` for interior mutability so it can be called through `&self`
+    /// (the memtable may already be inside an `Arc` when frozen). Safe to call
+    /// multiple times; only the first call has any effect.
+    pub fn freeze_range_tombstones(&self) {
+        if !self.range_tombstones.is_empty() {
+            self.immutable_range_tombstones.get_or_init(|| {
+                ImmutableRangeTombstoneSet::from_range_tombstone_set(&self.range_tombstones)
+            });
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/mem_table.rs:697]: Not fixed. The lazy OnceLock approach is idiomatic and the first-read latency is negligible for typical tombstone counts. Eagerly computing fragments during freeze_range_tombstones would shift cost to the write-critical path.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/lsm_storage.rs [hunk 23]
```diff
@@ -1600,7 +1620,41 @@ impl LsmStorageInner {
...
+        // Build merged range-tombstone fragments from all memtables.
+        // Active memtable: private per-scan fragments (no shared cache).
+        // Immutable memtables: shared cached fragments (borrowed, not cloned).
+        let range_ts_iter = if mvcc_read_ts.is_some() {
+            let active_frags = if !state.memtable.range_tombstones().is_empty() {
+                crate::range_tombstone::fragment_range(state.memtable.range_tombstones().raw())
+            } else {
+                Vec::new()
+            };
+            let mut lists: Vec<&[crate::range_tombstone::RangeTombstoneFragment]> = Vec::new();
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Fixed. Pre-allocated `lists` capacity with `Vec::with_capacity(1 + state.imm_memtables.len())`.

File: kv-engine/src/mem_table.rs [hunk 24]
```diff
@@ -23,6 +26,71 @@ use crate::{
...
+}
+
+impl ImmutableRangeTombstoneSet {
+    /// Wrap a `RangeTombstoneSet`'s raw skipmap into an immutable view.
+    pub fn from_range_tombstone_set(set: &RangeTombstoneSet) -> Self {
+        Self {
+            raw: Arc::clone(set.raw()),
+            fragments: OnceLock::new(),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/mem_table.rs:49]: Not fixed. Same reasoning as round 7 — the lazy OnceLock approach is idiomatic. Eagerly computing fragments during freeze would shift cost to the write-critical path with negligible read benefit for typical tombstone counts.

File: kv-engine/src/range_tombstone.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:454]: Not fixed. Same reasoning as rounds 4, 6, and 7 — the current sweep-line is efficient enough and the marginal improvement does not justify the added complexity.

File: kv-engine/src/range_tombstone.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:353]: Not fixed. Same reasoning as rounds 4, 6, and 7 — the O(F log F) sweep-line is already efficient. The event-based refactor adds complexity with diminishing returns for typical tombstone counts.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/range_tombstone.rs [hunk 25]
```diff
@@ -192,6 +193,280 @@ impl Default for RangeTombstoneSet {
...
+        }
+        // Binary search within covering_ts for the newest ts <= read_ts.
+        let ts_idx = frag.covering_ts.partition_point(|&ts| ts <= read_ts);
+        if ts_idx > 0 {
+            Some(frag.covering_ts[ts_idx - 1])
+        } else {
+            None
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:477]: Fixed. Extracted `find_newest_covering_ts` as a free function in the `range_tombstone` module. Both `RangeTombstoneIterator::newest_covering_ts` and `ImmutableRangeTombstoneSet::newest_covering_ts` now delegate to it.

File: kv-engine/src/range_tombstone.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:353]: Fixed. Simplified `starts_at` and `ends_at` from `Vec<Vec<(usize, u64)>>` to `Vec<Vec<u64>>` since the index was never used in the sweep-line — only the timestamp matters.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/range_tombstone.rs [hunk 26]
```diff
@@ -192,6 +193,289 @@ impl Default for RangeTombstoneSet {
...
+        let covering_ts: Vec<u64> = active.keys().copied().collect();
+
+        fragments.push(RangeTombstoneFragment {
+            start: frag_start,
+            end: frag_end,
+            covering_ts,
+        });
+    }
+
+    fragments
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:352]: Not fixed. This is the same suggestion declined in rounds 4, 6, 7, and 8. The current O(F log F) sweep-line is already efficient. The event-based refactor adds complexity with diminishing returns.

File: kv-engine/src/range_tombstone.rs [hunk 27]
```diff
@@ -192,6 +193,289 @@ impl Default for RangeTombstoneSet {
...
+
+        let covering_ts: Vec<u64> = active.keys().copied().collect();
+        fragments.push(RangeTombstoneFragment {
+            start: frag_start,
+            end: frag_end,
+            covering_ts,
+        });
+    }
+
+    fragments
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:453]: Not fixed. Same reasoning — this suggestion has been declined in multiple prior rounds. The current sweep-line is efficient enough.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/lsm_storage.rs [hunk 28]
```diff
@@ -1184,17 +1186,28 @@ impl LsmStorageInner {
         } else {
             key
         };
+        // Pre-compute the newest range tombstone ts once for both memtable and
+        // SST checks — avoids redundant scans of all memtables.
+        let range_ts = mvcc_read_ts.and_then(|rts| self.newest_memtable_range_ts(&state, key, rts));
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Fixed. Changed to `mvcc_read_ts.unwrap_or(u64::MAX)` so range tombstones are queried even in non-MVCC mode.

File: kv-engine/src/lsm_storage.rs [hunk 29]
```diff
@@ -1657,20 +1712,55 @@ impl LsmStorageInner {
         } else {
             key
         };
-        if let Some((val, kind, _key)) =
+        let range_ts = mvcc_read_ts.and_then(|rts| self.newest_memtable_range_ts(state, key, rts));
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Fixed. Same pattern — `mvcc_read_ts.unwrap_or(u64::MAX)` ensures range tombstones work in non-MVCC mode.

File: kv-engine/src/range_tombstone.rs [hunk 30]
```diff
@@ -192,6 +193,289 @@ impl Default for RangeTombstoneSet {
...
+
+        fragments.push(RangeTombstoneFragment {
+            start: frag_start,
+            end: frag_end,
+            covering_ts,
+        });
+    }
+
+    fragments
+}
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:353]: Not fixed. Same suggestion declined in rounds 4, 6, 7, 8, and 10. The current sweep-line is efficient enough.

File: kv-engine/src/range_tombstone.rs [hunk 31]
```diff
@@ -192,6 +193,289 @@ impl Default for RangeTombstoneSet {
...
+        let covering_ts: Vec<u64> = active.keys().copied().collect();
+        fragments.push(RangeTombstoneFragment {
+            start: frag_start,
+            end: frag_end,
+            covering_ts,
+        });
+    }
+
+    fragments
+}
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:454]: Not fixed. Same suggestion declined in multiple prior rounds.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/lsm_storage.rs [hunk 32]
```diff
@@ -1600,7 +1620,42 @@ impl LsmStorageInner {
...
+                None
+            } else {
+                let merged = crate::range_tombstone::merge_fragment_lists(&lists);
+                Some(crate::range_tombstone::RangeTombstoneIterator::new(
+                    Arc::from(merged),
+                ))
+            }
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Fixed. Added early short-circuit: compute `has_active` and `has_imm` before entering the block, skipping the entire fragment_range/merge logic when no range tombstones are present.

File: kv-engine/src/range_tombstone.rs [hunk 33]
```diff
@@ -192,6 +193,289 @@ impl Default for RangeTombstoneSet {
...
+    for (start, end, ts) in &tombstones {
+        let start_pos = endpoints
+            .binary_search(start)
+            .expect("start endpoint must exist in collected endpoints");
+        let end_pos = endpoints
+            .binary_search(end)
+            .expect("end endpoint must exist in collected endpoints");
+        starts_at[start_pos].push(*ts);
+        ends_at[end_pos].push(*ts);
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:295]: Not fixed. The intermediate tombstones vector is small and the current two-pass approach (collect then sweep) is clear and efficient. Iterating the SkipMap twice would add complexity with marginal benefit.

**Comment by ben1009** [comment]
All review comments addressed. Please re-review.
@coderabbitai @gemini-code-assist

File: kv-engine/src/range_tombstone.rs (hunk 30 above)
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:353]: Not fixed. This is the same suggestion declined in rounds 4, 6, 7, 8, 10, 11, and 12. The current sweep-line is efficient for the expected workload. No further action planned.

File: kv-engine/src/range_tombstone.rs (hunk 31 above)
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/range_tombstone.rs:454]: Not fixed. Same suggestion declined in multiple prior rounds. No further action planned.
